packages/core/contracts/l2/polygon/OracleBaseTunnel.sol

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "../../oracle/interfaces/FinderInterface.sol";
+import "../../oracle/implementation/Constants.sol";
+
+/**
+ * @notice Enforces lifecyle of price requests for deriving contract. Ensures that price request data is not duplicated
+ * and that only submitted price requests can be resolved.
+ */
+abstract contract OracleBaseTunnel {
+    enum RequestState { NeverRequested, Requested, Resolved }
+
+    struct Price {
+        RequestState state;
+        int256 price;
+    }
+
+    // Mapping of encoded price requests {identifier, time, ancillaryData} to Price objects.
+    mapping(bytes32 => Price) internal prices;
+
+    // Finder to provide addresses for DVM system contracts.
+    FinderInterface public finder;
+
+    event PriceRequestAdded(bytes32 indexed identifier, uint256 time, bytes ancillaryData);
+    event PushedPrice(bytes32 indexed identifier, uint256 time, bytes ancillaryData, int256 price);
+
+    /**
+     * @notice Constructor.
+     * @param _finderAddress finder to use to get addresses of DVM contracts.
+     */
+    constructor(address _finderAddress) {
+        finder = FinderInterface(_finderAddress);
+    }
+
+    /**
+     * @notice Enqueues a request (if a request isn't already present) for the given (identifier, time,
+     * ancillary data) combination. Will only emit an event if the request has never been requested.
+     */
+    function _requestPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) internal {
+        bytes32 priceRequestId = _encodePriceRequest(identifier, time, ancillaryData);
+        Price storage lookup = prices[priceRequestId];
+        if (lookup.state == RequestState.NeverRequested) {
+            lookup.state = RequestState.Requested;
+            emit PriceRequestAdded(identifier, time, ancillaryData);
+        }
+    }
+
+    /**
+     * @notice Publishes price for a requested query. Will only emit an event if the request has never been resolved.
+     */
+    function _publishPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData,
+        int256 price
+    ) internal {
+        bytes32 priceRequestId = _encodePriceRequest(identifier, time, ancillaryData);
+        Price storage lookup = prices[priceRequestId];
+        if (lookup.state == RequestState.Requested) {
```

Comment:
Note: Unlike in BeaconOracle, this _publishPrice method does NOT revert if called more than once. This is because we are not concerned about spam cross-chain messages.